ethash/src/compute.rs

```diff
@@ -467,159 +315,310 @@ fn calculate_dag_item(node_index: u32, cache: &[Node]) -> Node {
-	let cache_size = get_cache_size(block_number);
-
-	assert!(cache_size % NODE_BYTES == 0, "Unaligned cache size");
-	let num_nodes = cache_size / NODE_BYTES;
-
-	let mut nodes: Vec<Node> = Vec::with_capacity(num_nodes);
-	unsafe {
-		// Use uninit instead of unnecessarily writing `size_of::<Node>() * num_nodes` 0s
-		nodes.set_len(num_nodes);
-
-		sha3_512(&seedhash[0..32], &mut nodes.get_unchecked_mut(0).bytes);
-		for i in 1..num_nodes {
-			sha3::sha3_512(nodes.get_unchecked_mut(i).bytes.as_mut_ptr(), NODE_BYTES, nodes.get_unchecked(i - 1).bytes.as_ptr(), NODE_BYTES);
-		}
-
-		debug_assert_eq!(NODE_WORDS, 16);
-
-		// This _should_ get unrolled by the compiler, since it's not using the loop variable.
-		for _ in 0..ETHASH_CACHE_ROUNDS {
-			for i in 0..num_nodes {
-				let idx = *nodes.get_unchecked_mut(i).as_words().get_unchecked(0) as usize % num_nodes;
-				let mut data = nodes.get_unchecked((num_nodes - 1 + i) % num_nodes).clone();
-
-				unroll! {
-					for w in 0..16 {
-						*data.as_words_mut().get_unchecked_mut(w) ^= *nodes.get_unchecked(idx).as_words().get_unchecked(w);
-					}
-				}
-
-				sha3_512(&data.bytes, &mut nodes.get_unchecked_mut(i).bytes);
-			}
-		}
+#[cfg(test)]
+mod test {
+	use super::*;
+	use std::fs;
+
+	#[test]
+	fn test_get_cache_size() {
+		// https://github.com/ethereum/wiki/wiki/Ethash/ef6b93f9596746a088ea95d01ca2778be43ae68f#data-sizes
+		assert_eq!(16776896usize, get_cache_size(0));
+		assert_eq!(16776896usize, get_cache_size(1));
+		assert_eq!(16776896usize, get_cache_size(ETHASH_EPOCH_LENGTH - 1));
+		assert_eq!(16907456usize, get_cache_size(ETHASH_EPOCH_LENGTH));
+		assert_eq!(16907456usize, get_cache_size(ETHASH_EPOCH_LENGTH + 1));
+		assert_eq!(284950208usize, get_cache_size(2046 * ETHASH_EPOCH_LENGTH));
+		assert_eq!(285081536usize, get_cache_size(2047 * ETHASH_EPOCH_LENGTH));
+		assert_eq!(285081536usize, get_cache_size(2048 * ETHASH_EPOCH_LENGTH - 1));
 	}
 
-	Light {
-		block_number,
-		cache_dir: cache_dir.as_ref().to_path_buf(),
-		cache: nodes,
-		seed_compute: Mutex::new(seed_compute),
+	#[test]
+	fn test_get_data_size() {
+		// https://github.com/ethereum/wiki/wiki/Ethash/ef6b93f9596746a088ea95d01ca2778be43ae68f#data-sizes
+		assert_eq!(1073739904usize, get_data_size(0));
+		assert_eq!(1073739904usize, get_data_size(1));
+		assert_eq!(1073739904usize, get_data_size(ETHASH_EPOCH_LENGTH - 1));
+		assert_eq!(1082130304usize, get_data_size(ETHASH_EPOCH_LENGTH));
+		assert_eq!(1082130304usize, get_data_size(ETHASH_EPOCH_LENGTH + 1));
+		assert_eq!(18236833408usize, get_data_size(2046 * ETHASH_EPOCH_LENGTH));
+		assert_eq!(18245220736usize, get_data_size(2047 * ETHASH_EPOCH_LENGTH));
 	}
-}
 
-static CHARS: &'static [u8] = b"0123456789abcdef";
-fn to_hex(bytes: &[u8]) -> String {
-	let mut v = Vec::with_capacity(bytes.len() * 2);
-	for &byte in bytes.iter() {
-		v.push(CHARS[(byte >> 4) as usize]);
-		v.push(CHARS[(byte & 0xf) as usize]);
+	#[test]
+	fn test_difficulty_test() {
+		let hash = [
```

Comment:
This kind of formatting does not really make the code more readable. Could you please revert it?

Reply:
Sorry, I use rustfmt because I don't like manually formatting. I didn't mean to commit this at all. The next version of rustfmt will support the visual block formatting that we had before (so first element on the next line and tabbed in, consecutive elements on same line). Thanks for catching this.